src/components/hooks/useOnNetworkReconnect.js

```diff
@@ -0,0 +1,21 @@
+import {useRef, useContext, useEffect} from 'react';
+import {NetworkContext} from '../OnyxProvider';
+
+export default function useOnNetworkReconnect(callback) {
+    const {isOffline} = useContext(NetworkContext);
+    const prevOfflineStatusRef = useRef(isOffline);
+    useEffect(() => {
+        // If we were offline before and now we are not offline then we just reconnected
+        const didReconnect = prevOfflineStatusRef.current && !isOffline;
+        if (!didReconnect) {
+            return;
+        }
+
+        callback();
```

Comment:
Can we not call the callback with the `didReconnect` so we can do further optimization like @kidroca mentioned in the [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1681400895336669?thread_ts=1681117333.835879&cid=C01GTK53T8Q)?

Reply:
It looks like the custom hook here is meant only if the network reconnects but maybe we just need `useNetworkState` or something? And then callback can have the updated state of the network and decide what they need to do next.

Reply:
I think this is fine, the idea is to cleanup and simplify, so we have 1 `if` statement here instead of many `if` statements everywhere we need to run such checks 

In terms of usage we achieve a declarative approach
1. on reconnect
2. do something 

No ifs

Reply:
A minor problem here is that `callback` would trigger the hook to re-run on-rerender, when it is passed inline 
For this particular hook this won't cause anything unexpected, just triggering the effect more often than necessary 

#### Problem
We have a custom effect hook that triggers an `action` (callback)
We don't want the `action` itself to be a reason to run/re-run the custom effect
The custom effect should only run the action when a list of dependencies change
But the action itself shouldn't be a dependency

#### Proposal

_Marc, I think you know my way of dealing with that:_

Exclude the action from dependencies by capturing it to a ref (internal hook logic)
This way we only run the effect when network conditions change
But we still use the latest value of the action when we do need to run it

```js
// prevent the callback param from re-triggering the hook if it changes
// we want to only trigger the hook on reconnect, and then use the latest value for the 
// `callback` so we keep a ref updated with the latest value
const callbackRef = useRef(callback);
callbackRef.current = callback;

useEffect(() => {
  // ...
  callbackRef.current();
}, [isOffline])
```

Reply:
Ah so, one curiosity I have in response is if it matters that the effect re-runs when the callback changes? Since we will only do the callback if the conditions of the props are correct i.e. the `isOffline` changed in the right way. It would almost be a no-op - but maybe not entirely true that it does _nothing_ (still has to do the comparison).

I like the suggestion to use the ref anyway so I think we can do it! But mostly just curious if there was some worse consequence to having the effect run more than necessary.

Reply:
I have these 2 concerns. Though I can't measure their possible impact right now

**Performacnewise** triggering the effect would happen on each render unless the caller uses a `useCallback` wrapping for the callback, or it's a static callback (like in the Modal file change)
If we only have a few usages/instances of the hook I wouldn't think it's a big deal, but since it might be used on a frequently rendered component like Avatar it multiplies the potential issue 
It probably won't be noticeable, but little things add up 
_(the actual call to the function is the overhead even if exists immediately - it has to be orchestrated stacked/destaced, at least that's what I think, I don't have a quote to reference)_ 

By using ref for the callback we don't have to convey to developers there might be performance implications so that they should use `useCallback` when passing params to the hook

**Code pattern wise**  someone is bound to need to write a new hook and might use this one as inspiration, not realizing that adding the `callback` to dependencies here does not affect the effect, but it might be a problem for their use case. They'll probably find out soon that there's a problem (the problem stated above) and look for a solution (similar to the proposed solution) - so by doing the ref pattern here we've saved them, but they'll never know 😆

Reply:
I think they're great points all around. Thanks for your guidance here 🙇